File: sdk/java-holons/src/test/java/org/organicprogramming/holons/DiscoverTests.java
package org.organicprogramming.holons;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.organicprogramming.holons.DiscoveryTestSupport.PackageSeed;
import org.organicprogramming.holons.DiscoveryTestSupport.RuntimeFixture;
import org.organicprogramming.holons.DiscoveryTypes.DiscoverResult;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.organicprogramming.holons.DiscoveryTestSupport.holonInfo;
import static org.organicprogramming.holons.DiscoveryTestSupport.holonRef;
import static org.organicprogramming.holons.DiscoveryTestSupport.runtimeEnv;
import static org.organicprogramming.holons.DiscoveryTestSupport.runtimeFixture;
import static org.organicprogramming.holons.DiscoveryTestSupport.slugs;
import static org.organicprogramming.holons.DiscoveryTestSupport.writePackageHolon;

class DiscoverTests {

    @AfterEach
    void resetHooks() {
        Discover.setPackageProbeForTests(null);
        Discover.setSourceBridgeForTests(null);
        Discover.setBundleRootResolverForTests(null);
    }

    @Test
    void discoverAllLayers(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        Path bundleRoot = root.resolve("TestApp.app").resolve("Contents").resolve("Resources").resolve("Holons");

        writePackageHolon(bundleRoot.resolve("bundle.holon"),
                new PackageSeed("uuid-bundle", "Bundle", "Holon").slug("bundle"));
        writePackageHolon(root.resolve("cwd-alpha.holon"),
                new PackageSeed("uuid-cwd-alpha", "Cwd", "Alpha").slug("cwd-alpha"));
        writePackageHolon(root.resolve(".op").resolve("build").resolve("built-beta.holon"),
                new PackageSeed("uuid-built-beta", "Built", "Beta").slug("built-beta"));
        writePackageHolon(fixture.opBin().resolve("installed-gamma.holon"),
                new PackageSeed("uuid-installed-gamma", "Installed", "Gamma").slug("installed-gamma"));
        writePackageHolon(fixture.opHome().resolve("cache").resolve("deps").resolve("cached-delta.holon"),
                new PackageSeed("uuid-cached-delta", "Cached", "Delta").slug("cached-delta"));

        Discover.setBundleRootResolverForTests(() -> bundleRoot);
        Discover.setSourceBridgeForTests((expression, searchRoot, limit, timeout) -> {
            DiscoverResult bridged = new DiscoverResult();
            bridged.found.add(holonRef(root.resolve("source-epsilon"), "source-epsilon", "uuid-source-epsilon", "Source", "Epsilon"));
            return bridged;
        });

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.ALL, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("built-beta", "bundle", "cached-delta", "cwd-alpha", "installed-gamma", "source-epsilon"), slugs(result));
        }
    }

    @Test
    void filterBySpecifiers(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("cwd-alpha.holon"),
                new PackageSeed("uuid-cwd-alpha", "Cwd", "Alpha").slug("cwd-alpha"));
        writePackageHolon(root.resolve(".op").resolve("build").resolve("built-beta.holon"),
                new PackageSeed("uuid-built-beta", "Built", "Beta").slug("built-beta"));
        writePackageHolon(fixture.opBin().resolve("installed-gamma.holon"),
                new PackageSeed("uuid-installed-gamma", "Installed", "Gamma").slug("installed-gamma"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(
                    Discover.LOCAL,
                    null,
                    root.toString(),
                    Discover.BUILT | Discover.INSTALLED,
                    Discover.NO_LIMIT,
                    Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("built-beta", "installed-gamma"), slugs(result));
        }
    }

    @Test
    void matchBySlug(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));
        writePackageHolon(root.resolve("beta.holon"), new PackageSeed("uuid-beta", "Beta", "Two").slug("beta"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, "beta", root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("beta"), slugs(result));
        }
    }

    @Test
    void matchByAlias(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"),
                new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha").aliases(List.of("first")));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, "first", root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("alpha"), slugs(result));
        }
    }

    @Test
    void matchByUuidPrefix(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"),
                new PackageSeed("12345678-aaaa-bbbb-cccc-dddddddddddd", "Alpha", "One").slug("alpha"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, "12345678", root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("alpha"), slugs(result));
        }
    }

    @Test
    void matchByPath(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        Path alpha = root.resolve("alpha.holon");
        writePackageHolon(alpha, new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, alpha.toString(), root.toString(), Discover.ALL, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("alpha"), slugs(result));
            assertEquals(alpha.toAbsolutePath().normalize().toUri().toString(), result.found.get(0).url);
        }
    }

    @Test
    void limitOne(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));
        writePackageHolon(root.resolve("beta.holon"), new PackageSeed("uuid-beta", "Beta", "Two").slug("beta"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CWD, 1, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(1, result.found.size());
        }
    }

    @Test
    void limitZeroMeansUnlimited(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));
        writePackageHolon(root.resolve("beta.holon"), new PackageSeed("uuid-beta", "Beta", "Two").slug("beta"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CWD, 0, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(2, result.found.size());
        }
    }

    @Test
    void negativeLimitReturnsEmpty(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CWD, -1, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertTrue(result.found.isEmpty());
        }
    }

    @Test
    void invalidSpecifiers(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), 0xFF, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertTrue(result.error.contains("invalid specifiers"));
        }
    }

    @Test
    void specifiersZeroTreatedAsAll(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("cwd-alpha.holon"),
                new PackageSeed("uuid-cwd-alpha", "Cwd", "Alpha").slug("cwd-alpha"));
        writePackageHolon(root.resolve(".op").resolve("build").resolve("built-beta.holon"),
                new PackageSeed("uuid-built-beta", "Built", "Beta").slug("built-beta"));
        writePackageHolon(fixture.opBin().resolve("installed-gamma.holon"),
                new PackageSeed("uuid-installed-gamma", "Installed", "Gamma").slug("installed-gamma"));
        writePackageHolon(fixture.opHome().resolve("cache").resolve("cached-delta.holon"),
                new PackageSeed("uuid-cached-delta", "Cached", "Delta").slug("cached-delta"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult all = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.ALL, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            DiscoverResult zero = Discover.Discover(Discover.LOCAL, null, root.toString(), 0, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", all.error);
            assertEquals("", zero.error);
            assertEquals(slugs(all), slugs(zero));
        }
    }

    @Test
    void nullExpressionReturnsAll(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));
        writePackageHolon(root.resolve("beta.holon"), new PackageSeed("uuid-beta", "Beta", "Two").slug("beta"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(2, result.found.size());
        }
    }

    @Test
    void missingExpressionReturnsEmpty(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, "missing", root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertTrue(result.found.isEmpty());
        }
    }

    @Test
    void excludedDirsSkipped(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("kept.holon"), new PackageSeed("uuid-kept", "Kept", "Holon").slug("kept"));
        for (String skipped : List.of(".git", ".op", "node_modules", "vendor", "build", "testdata", ".cache")) {
            writePackageHolon(root.resolve(skipped).resolve("hidden.holon"),
                    new PackageSeed("uuid-" + skipped, "Ignored", "Holon").slug("ignored-" + skipped.replace('.', '_')));
        }

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("kept"), slugs(result));
        }
    }

    @Test
    void deduplicateByUuid(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        Path cwd = root.resolve("alpha.holon");
        writePackageHolon(cwd, new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));
        writePackageHolon(root.resolve(".op").resolve("build").resolve("alpha-built.holon"),
                new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha-built"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.ALL, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(1, result.found.size());
            assertEquals(cwd.toAbsolutePath().normalize().toUri().toString(), result.found.get(0).url);
        }
    }

    @Test
    void holonJsonFastPath(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        AtomicInteger probeCalls = new AtomicInteger();
        Discover.setPackageProbeForTests(packageDir -> {
            probeCalls.incrementAndGet();
            return holonInfo("alpha", "uuid-alpha", "Alpha", "One");
        });
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("alpha"), slugs(result));
            assertEquals(0, probeCalls.get());
        }
    }

    @Test
    void describeFallbackWhenHolonJsonMissing(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        Path alpha = root.resolve("alpha.holon");
        Files.createDirectories(alpha);
        AtomicInteger probeCalls = new AtomicInteger();
        Discover.setPackageProbeForTests(packageDir -> {
            probeCalls.incrementAndGet();
            assertEquals(alpha.toAbsolutePath().normalize(), packageDir);
            return holonInfo("alpha", "uuid-alpha", "Alpha", "One");
        });

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("alpha"), slugs(result));
            assertEquals(1, probeCalls.get());
        }
    }

    @Test
    void siblingsLayer(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        Path bundleRoot = root.resolve("App.app").resolve("Contents").resolve("Resources").resolve("Holons");
        writePackageHolon(bundleRoot.resolve("bundle.holon"),
                new PackageSeed("uuid-bundle", "Bundle", "Holon").slug("bundle"));
        Discover.setBundleRootResolverForTests(() -> bundleRoot);

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.SIBLINGS, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("bundle"), slugs(result));
        }
    }

    @Test
    void sourceLayerOffloadsToLocalOp(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        AtomicInteger calls = new AtomicInteger();
        AtomicReference<String> capturedExpression = new AtomicReference<>("unset");
        AtomicReference<Path> capturedRoot = new AtomicReference<>();
        Discover.setSourceBridgeForTests((expression, searchRoot, limit, timeout) -> {
            calls.incrementAndGet();
            capturedExpression.set(expression);
            capturedRoot.set(searchRoot);
            DiscoverResult bridged = new DiscoverResult();
            bridged.found.add(holonRef(root.resolve("proto-holon"), "proto-holon", "uuid-proto", "Proto", "Holon"));
            return bridged;
        });

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.SOURCE, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("proto-holon"), slugs(result));
            assertEquals(1, calls.get());
            assertEquals(null, capturedExpression.get());
            assertEquals(root.toAbsolutePath().normalize(), capturedRoot.get());
        }
    }

    @Test
    void builtLayer(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve(".op").resolve("build").resolve("built.holon"),
                new PackageSeed("uuid-built", "Built", "Holon").slug("built"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.BUILT, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("built"), slugs(result));
        }
    }

    @Test
    void installedLayer(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(fixture.opBin().resolve("installed.holon"),
                new PackageSeed("uuid-installed", "Installed", "Holon").slug("installed"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.INSTALLED, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("installed"), slugs(result));
        }
    }

    @Test
    void cachedLayer(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(fixture.opHome().resolve("cache").resolve("deep").resolve("cached.holon"),
                new PackageSeed("uuid-cached", "Cached", "Holon").slug("cached"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, root.toString(), Discover.CACHED, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("cached"), slugs(result));
        }
    }

    @Test
    void nilRootDefaultsToCwd(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        writePackageHolon(root.resolve("alpha.holon"), new PackageSeed("uuid-alpha", "Alpha", "One").slug("alpha"));

        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, null, Discover.CWD, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertEquals("", result.error);
            assertEquals(List.of("alpha"), slugs(result));
        }
    }

    @Test
    void emptyRootReturnsError(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        try (var env = runtimeEnv(fixture)) {
            DiscoverResult result = Discover.Discover(Discover.LOCAL, null, "", Discover.ALL, Discover.NO_LIMIT, Discover.NO_TIMEOUT);
            assertTrue(result.error.contains("root cannot be empty"));
        }
    }

    @Test
    void unsupportedScopeReturnsError(@TempDir Path root) throws Exception {
        RuntimeFixture fixture = runtimeFixture(root);
        try (var env = runtimeEnv(fixture)) {
            assertNotEquals("", Discover.Discover(Discover.PROXY, null, root.toString(), Discover.ALL, Discover.NO_LIMIT, Discover.NO_TIMEOUT).error);
            assertNotEquals("", Discover.Discover(Discover.DELEGATED, null, root.toString(), Discover.ALL, Discover.NO_LIMIT, Discover.NO_TIMEOUT).error);
        }
    }
}

File: sdk/kotlin-holons/src/test/kotlin/org/organicprogramming/holons/DiscoverTests.kt
package org.organicprogramming.holons

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class DiscoverTests {
    @Test
    fun discoverAllLayers() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("cwd-alpha.holon"), PackageSeed("cwd-alpha", "uuid-cwd-alpha", "Cwd", "Alpha", entrypoint = "cwd-alpha"))
        writePackageHolon(fixture.root.resolve(".op/build/built-beta.holon"), PackageSeed("built-beta", "uuid-built-beta", "Built", "Beta", entrypoint = "built-beta"))
        writePackageHolon(fixture.opBin.resolve("installed-gamma.holon"), PackageSeed("installed-gamma", "uuid-installed-gamma", "Installed", "Gamma", entrypoint = "installed-gamma"))
        writePackageHolon(fixture.opHome.resolve("cache/deps/cached-delta.holon"), PackageSeed("cached-delta", "uuid-cached-delta", "Cached", "Delta", entrypoint = "cached-delta"))

        val result = Discover(LOCAL, null, fixture.root.toString(), ALL, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("built-beta", "cached-delta", "cwd-alpha", "installed-gamma"), sortedSlugs(result))
    }

    @Test
    fun filterBySpecifiers() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("cwd-alpha.holon"), PackageSeed("cwd-alpha", "uuid-cwd-alpha", "Cwd", "Alpha", entrypoint = "cwd-alpha"))
        writePackageHolon(fixture.root.resolve(".op/build/built-beta.holon"), PackageSeed("built-beta", "uuid-built-beta", "Built", "Beta", entrypoint = "built-beta"))
        writePackageHolon(fixture.opBin.resolve("installed-gamma.holon"), PackageSeed("installed-gamma", "uuid-installed-gamma", "Installed", "Gamma", entrypoint = "installed-gamma"))

        val result = Discover(LOCAL, null, fixture.root.toString(), BUILT or INSTALLED, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("built-beta", "installed-gamma"), sortedSlugs(result))
    }

    @Test
    fun matchBySlug() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))
        writePackageHolon(fixture.root.resolve("beta.holon"), PackageSeed("beta", "uuid-beta", "Beta", "Two", entrypoint = "beta"))

        val result = Discover(LOCAL, "beta", fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("beta"), sortedSlugs(result))
    }

    @Test
    fun matchByAlias() = withRuntimeFixture { fixture ->
        writePackageHolon(
            fixture.root.resolve("alpha.holon"),
            PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha", aliases = listOf("first")),
        )

        val result = Discover(LOCAL, "first", fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("alpha"), sortedSlugs(result))
    }

    @Test
    fun matchByUuidPrefix() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "12345678-aaaa", "Alpha", "One", entrypoint = "alpha"))

        val result = Discover(LOCAL, "12345678", fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("alpha"), sortedSlugs(result))
    }

    @Test
    fun matchByPath() = withRuntimeFixture { fixture ->
        val target = fixture.root.resolve("beta.holon")
        writePackageHolon(target, PackageSeed("beta", "uuid-beta", "Beta", "Two", entrypoint = "beta"))

        val result = Discover(LOCAL, target.toString(), fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(1, result.found.size)
        assertEquals(fileURL(target), result.found.single().url)
    }

    @Test
    fun limitOne() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))
        writePackageHolon(fixture.root.resolve("beta.holon"), PackageSeed("beta", "uuid-beta", "Beta", "Two", entrypoint = "beta"))

        val result = Discover(LOCAL, null, fixture.root.toString(), CWD, 1, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(1, result.found.size)
    }

    @Test
    fun limitZeroMeansUnlimited() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))
        writePackageHolon(fixture.root.resolve("beta.holon"), PackageSeed("beta", "uuid-beta", "Beta", "Two", entrypoint = "beta"))

        val result = Discover(LOCAL, null, fixture.root.toString(), CWD, 0, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(2, result.found.size)
    }

    @Test
    fun negativeLimitReturnsEmpty() = withRuntimeFixture { fixture ->
        val result = Discover(LOCAL, null, fixture.root.toString(), CWD, -1, NO_TIMEOUT)
        assertNull(result.error)
        assertTrue(result.found.isEmpty())
    }

    @Test
    fun invalidSpecifiers() = withRuntimeFixture { fixture ->
        val result = Discover(LOCAL, null, fixture.root.toString(), 0xFF, NO_LIMIT, NO_TIMEOUT)
        assertNotNull(result.error)
        assertTrue(result.error!!.contains("invalid specifiers"))
    }

    @Test
    fun specifiersZeroTreatedAsAll() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("cwd-alpha.holon"), PackageSeed("cwd-alpha", "uuid-cwd-alpha", "Cwd", "Alpha", entrypoint = "cwd-alpha"))
        writePackageHolon(fixture.root.resolve(".op/build/built-beta.holon"), PackageSeed("built-beta", "uuid-built-beta", "Built", "Beta", entrypoint = "built-beta"))
        writePackageHolon(fixture.opBin.resolve("installed-gamma.holon"), PackageSeed("installed-gamma", "uuid-installed-gamma", "Installed", "Gamma", entrypoint = "installed-gamma"))
        writePackageHolon(fixture.opHome.resolve("cache/deps/cached-delta.holon"), PackageSeed("cached-delta", "uuid-cached-delta", "Cached", "Delta", entrypoint = "cached-delta"))

        val allResult = Discover(LOCAL, null, fixture.root.toString(), ALL, NO_LIMIT, NO_TIMEOUT)
        val zeroResult = Discover(LOCAL, null, fixture.root.toString(), 0, NO_LIMIT, NO_TIMEOUT)

        assertNull(allResult.error)
        assertNull(zeroResult.error)
        assertEquals(sortedSlugs(allResult), sortedSlugs(zeroResult))
    }

    @Test
    fun nullExpressionReturnsAll() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))
        writePackageHolon(fixture.root.resolve("beta.holon"), PackageSeed("beta", "uuid-beta", "Beta", "Two", entrypoint = "beta"))

        val result = Discover(LOCAL, null, fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(2, result.found.size)
    }

    @Test
    fun missingExpressionReturnsEmpty() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))

        val result = Discover(LOCAL, "", fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertTrue(result.found.isEmpty())
    }

    @Test
    fun excludedDirsSkipped() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("kept/alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))
        listOf(
            ".git/hidden.holon",
            ".op/hidden.holon",
            "node_modules/hidden.holon",
            "vendor/hidden.holon",
            "build/hidden.holon",
            "testdata/hidden.holon",
            ".cache/hidden.holon",
        ).forEachIndexed { index, relative ->
            writePackageHolon(
                fixture.root.resolve(relative),
                PackageSeed("ignored-$index", "uuid-ignored-$index", "Ignored", "Holon", entrypoint = "ignored-$index"),
            )
        }

        val result = Discover(LOCAL, null, fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("alpha"), sortedSlugs(result))
    }

    @Test
    fun deduplicateByUuid() = withRuntimeFixture { fixture ->
        val cwdPath = fixture.root.resolve("alpha.holon")
        val builtPath = fixture.root.resolve(".op/build/alpha-built.holon")
        writePackageHolon(cwdPath, PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))
        writePackageHolon(builtPath, PackageSeed("alpha-built", "uuid-alpha", "Alpha", "One", entrypoint = "alpha-built"))

        val result = Discover(LOCAL, null, fixture.root.toString(), ALL, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(1, result.found.size)
        assertEquals(fileURL(cwdPath), result.found.single().url)
    }

    @Test
    fun holonJsonFastPath() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))

        var probeCalls = 0
        packageDescribeProbeOverride = { dir, _ ->
            probeCalls += 1
            HolonRef(url = fileURL(dir), error = "probe should not run")
        }

        val result = Discover(LOCAL, null, fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(0, probeCalls)
        assertEquals(listOf("alpha"), sortedSlugs(result))
    }

    @Test
    fun describeFallbackWhenHolonJsonMissing() = withRuntimeFixture { fixture ->
        val pkg = fixture.root.resolve("static-only.holon")
        java.nio.file.Files.createDirectories(pkg)
        var probeCalls = 0
        packageDescribeProbeOverride = { dir, _ ->
            probeCalls += 1
            HolonRef(
                url = fileURL(dir),
                info = HolonInfo(
                    slug = "static-only",
                    uuid = "static-describe-uuid",
                    identity = IdentityInfo("Static", "Only", "Registered at startup."),
                    lang = "kotlin",
                    runner = "shell",
                    status = "draft",
                    kind = "native",
                    transport = "",
                    entrypoint = "static-wrapper",
                    architectures = emptyList(),
                    hasDist = true,
                    hasSource = false,
                ),
            )
        }

        val result = Discover(LOCAL, null, fixture.root.toString(), CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(1, probeCalls)
        assertEquals(listOf("static-only"), sortedSlugs(result))
    }

    @Test
    fun siblingsLayer() = withRuntimeFixture { fixture ->
        val siblingsRoot = fixture.root.resolve("bundle/Holons")
        System.setProperty("holons.siblings.root", siblingsRoot.toString())
        writePackageHolon(siblingsRoot.resolve("bundle.holon"), PackageSeed("bundle", "uuid-bundle", "Bundle", "Holon", entrypoint = "bundle"))

        val result = Discover(LOCAL, null, fixture.root.toString(), SIBLINGS, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("bundle"), sortedSlugs(result))
    }

    @Test
    fun sourceLayerOffloadsToLocalOp() = withRuntimeFixture { fixture ->
        val sourceDir = fixture.root.resolve("proto-holon")
        var calls = 0
        sourceDiscoverOffload = { scope, expression, root, specifiers, limit, timeout ->
            calls += 1
            assertEquals(LOCAL, scope)
            assertNull(expression)
            assertEquals(fixture.root.toString(), root)
            assertEquals(SOURCE, specifiers)
            assertEquals(NO_LIMIT, limit)
            assertEquals(3210, timeout)
            DiscoverResult(
                found = listOf(
                    HolonRef(
                        url = fileURL(sourceDir),
                        info = HolonInfo(
                            slug = "proto-holon",
                            uuid = "uuid-proto",
                            identity = IdentityInfo("Proto", "Holon"),
                            lang = "kotlin",
                            runner = "shell",
                            status = "draft",
                            kind = "service",
                            transport = "",
                            entrypoint = "proto-holon",
                            architectures = emptyList(),
                            hasDist = false,
                            hasSource = true,
                        ),
                    ),
                ),
            )
        }

        val result = Discover(LOCAL, null, fixture.root.toString(), SOURCE, NO_LIMIT, 3210)
        assertNull(result.error)
        assertEquals(1, calls)
        assertEquals(listOf("proto-holon"), sortedSlugs(result))
    }

    @Test
    fun builtLayer() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve(".op/build/built.holon"), PackageSeed("built", "uuid-built", "Built", "Holon", entrypoint = "built"))

        val result = Discover(LOCAL, null, fixture.root.toString(), BUILT, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("built"), sortedSlugs(result))
    }

    @Test
    fun installedLayer() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.opBin.resolve("installed.holon"), PackageSeed("installed", "uuid-installed", "Installed", "Holon", entrypoint = "installed"))

        val result = Discover(LOCAL, null, fixture.root.toString(), INSTALLED, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("installed"), sortedSlugs(result))
    }

    @Test
    fun cachedLayer() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.opHome.resolve("cache/deep/cached.holon"), PackageSeed("cached", "uuid-cached", "Cached", "Holon", entrypoint = "cached"))

        val result = Discover(LOCAL, null, fixture.root.toString(), CACHED, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("cached"), sortedSlugs(result))
    }

    @Test
    fun nilRootDefaultsToCwd() = withRuntimeFixture { fixture ->
        writePackageHolon(fixture.root.resolve("alpha.holon"), PackageSeed("alpha", "uuid-alpha", "Alpha", "One", entrypoint = "alpha"))

        val result = Discover(LOCAL, null, null, CWD, NO_LIMIT, NO_TIMEOUT)
        assertNull(result.error)
        assertEquals(listOf("alpha"), sortedSlugs(result))
    }

    @Test
    fun emptyRootReturnsError() = withRuntimeFixture {
        val result = Discover(LOCAL, null, "", ALL, NO_LIMIT, NO_TIMEOUT)
        assertNotNull(result.error)
        assertTrue(result.error!!.contains("root"))
    }

    @Test
    fun unsupportedScopeReturnsError() = withRuntimeFixture { fixture ->
        val proxy = Discover(PROXY, null, fixture.root.toString(), ALL, NO_LIMIT, NO_TIMEOUT)
        val delegated = Discover(DELEGATED, null, fixture.root.toString(), ALL, NO_LIMIT, NO_TIMEOUT)
        assertNotNull(proxy.error)
        assertNotNull(delegated.error)
    }
}
